### ISSUE TITLE
New Windows Installer Testing

### ISSUE THREAD

**Issue by Moburma** [swfans/swars#79]
As you may have noticed, the official Windows installer for the port is now years old, and doesn't support the latest developments (i.e. the change to how language files work). It also doesn't support what is now the most common version of the game (the GOG installer), and doesn't support the German CD release, either.

As such I've been working on updating it and making a newer, nicer and more functional installer. The original one was a good starting point, but now I'm quite confident that the latest version should cover nearly all bases from which you may be trying to install the game from. There's also an actual proper uninstaller now, too.

I've done lots of testing, but of course with these things you never really know they're 100% until they are out in the wild for people to really test them. If you would like to help test this new build, you can download the latest github actions build packaged up at the following link:

https://uploadnow.io/f/SjmGR0r

Obvious caveats apply - don't install it over your existing game if you don't want to risk losing anything, and be prepared for some possibly horrendous bug that I may have somehow missed. Please try it out and report success or failure. There are a couple of minor issues that are flagged in the installer itself, but I will not mention them here to see if what I have put in so far is good enough to mitigate them or if more is needed.

Also let me know of any other functionality that might be useful (like maybe uninstall while preserving saved game directory only?).

**Comment by mefistotelis** [comment]
Hm, my VirtualBox has trouble mounting an image with cd audio.. but tested without CDA, and it worked ok.

**Comment by Moburma** [comment]
Yes, that is something I was alluding to above. I did put in a warning if you hover over the game music option in the components page, but I think you've confirmed it's too easily missed, so I will put in a clearer warning. I think it's another common use case - I actually had this problem waaay back in the day when the original release of the port came out and was confused then.
Unfortunately I don't think there's any way to rip from a mounted iso, the workaround is simply not to do it in the installer and tell the user to do it themselves.

![iso issue](https://i.ibb.co/P4Hf2L2/isoinstall.png)

**Comment by mefistotelis** [comment]
>  I did put in a warning if you hover over the game music option in the components page, but I think you've confirmed it's too easily missed

I didn't really missed it, I just assumed that VirtualBox, being a simulator detached from any HW, will support the format properly. But it looks like it does not, even though multi-track CUE/BIN support was added back in version 5.2 (at least, this is what they claim in release notes).

**Comment by Moburma** [comment]
> seems to work without any problems.
> 
> side note: the default key "backslash" for grouping soldiers is not found in any of the normal keys on scandinavian keyboard layout.

Thanks, out of interest what version did you install, GOG? 

The keyboard thing is off topic, but I actually get the same problem with my UK keyboard, it's actually some non-standard keyboard I have and a few games don't recognise that key (DCS world is another).

**Comment by goofyseeker311** [comment]
gog version.

**Comment by goofyseeker311** [comment]
why does the game or installer put a win.gif/.png on my d:\ root?

**Comment by Moburma** [comment]
Oops, I was using that while testing something early on, was pretty certain I'd reverted it, it's not set there in my latest version. Thanks for finding it, this is exactly the sort of thing I wanted to squash before publishing.
I've uploaded a new version here, this doesn't put the image there for me:  https://uploadnow.io/files/NSkpK8M

**Comment by Moburma** [comment]
Another new version. This won't be very interesting for most people, but I've now moved over to my new more generic Gogisoripper tool instead of the original Syndicate Wars specific swisoripper tool for ripping the GOG audio tracks. What is slightly more interesting is I realised the GOG .cue file is actually wrong and the audio tracks are slightly misaligned. This is most obvious if you play track 3 of the GOG version in-game, which is missing the first second or two of music. As such I've gone back and rechecked the track boundaries by hand against a rip from a real CD and now the versions ripped by this installer are correct and don't miss any audio/include bits from the previous track like the GOG release does.

Link: https://uploadnow.io/f/FG0F31r

**Comment by mefistotelis** [comment]
The old `libpng3` is no longer necessary.

**Comment by Moburma** [comment]
One thing I noticed as well post-pull-request is there are duplicate lines at line 376 and 377, 380 and 381. Goto langncd_store only needs calling once.

**Comment by mefistotelis** [comment]
Will fix as well.

**Comment by Moburma** [comment]
More a note to myself, but one mildly annoying problem is that the progress bar never goes to 100% at the end of install due to the number of Return commands used in the functions. Nsis considers this "missed" progress. Really needs a re-write of the old Vexillum installer stuff to fix it (or an ugly workaround forcing the progress), not sure if it's worth it though.

**Comment by mefistotelis** [comment]
We should prepare a "release", as people are installing a version from April and asking strange questions...
Could you prepare a release notes for the v0.3.5?
Also, would be nice if the installer could "!include" the header file with version number, instead of having a version within the main script.

**Comment by Moburma** [comment]
Yes, I was just about to suggest the same. I'll put together release notes (be nice to have these displayed in the installer as a page, yes?) and try and get the version header working, I saw you started on that. I might try and fix the 100% progress issue too, saw a few ways to do it.

**Comment by mefistotelis** [comment]
> be nice to have these displayed in the installer as a page, yes?

My main intention is to paste them to the release message, as there were 700 commits since previous release. Though you can add it to the installer as well, up to you.

> try and get the version header working

It is in the same folder where your NSI script is copied, so should go without any issues.
(the automatic builds run the installer creation from `release/res` folder)

**Comment by Moburma** [comment]
How about this for release notes:

Release XXX

Improvements since release 0.3.3.797:

General Game Improvements

- MIDI tension music is now supported using the WildMIDI software MIDI library
- Player configuration preferences are now saved along with game saves. These work based on the login name the player chooses, and will preserve the menu settings between game sessions. Therefore the game will remember settings such as custom control bindings, high resolution/low resolution, Deep Radar and MIDI tension/CD track between sessions as long as the same login name is used, and they no longer have to be set again each time on game load.
- As a corollary to the above, the login name can now be set at the command line using the switch -u <username>. e.g. "swars.exe -u James" will load the preferences set for the user James. This way it's possible to start levels from the command line and have options such as Deep Radar and custom key controls already set.
- Campaigns have now been separated out into discrete files, instead of using the original ALL000.MIS file. This means there can now be per-campaign settings, and all three game campaigns no longer have to exist within the cramped 120 mission limit the original game had. Players should notice no difference, but mission objectives and campaign level orders can now be modified by altering the corresponding .ini file in the config directory in a text editor. Eurocorp campaign: miss000.ini, Church of the New Epoch campaign: miss001.ini, Unguided Campaign: miss002.ini. There is also a fourth campaign miss003.ini that contains all other cut levels for the game.
- As a result of the above, loading levels from the command line now uses a different syntax: -m <campaign number>,<mission index number>. For example, to play the first mission of the Church of the New Epoch campaign, you would use "swars.exe -m 1,48". To play the last level of the Eurocorp campaign, "swars.exe -m 0,101" would be used.
- City data has now been moved to cities.ini in the config directory. This allows modification of existing cities and new cities to be defined on the map screen
- Cybernetic Modification data has been moved to cybmods.ini in the config directory. This allows modification of e.g. research cost/time, and purchase costs.
- Weapon data has been moved to weapons.ini in the config directory. This allows modification of weapon stats, including weapon damage, accuracy, research cost/time and purchase cost.
- People data has been moved to people.ini in the config directory. This allows modification of baseline stats of game characters, including Health, speed, and Persuaderton cost/power. Note that many people stats are set explicitly per-level Thing and will override what is set here.
- Pre-alpha level file formats are now natively supported. This means old cut levels using level formats 9 - 12 that previously could not be loaded without being first converted by a third-party tool can now be played. Examples are Unguided Missions 67 and 70
- In-game text now uses per-campaign alltext.wad/idx files. They are also now separated by language, meaning it's possible to change between game language by changing the language set in the config.ini file (note, sound and FMV files would also need to be changed to match). Text is now stored in PO files in the source, making it easier to create translations of all text.

Unguided Campaign Changes

- Unguided Campaign now available from the Login screen. Note this is a first draft and unfinished and simply presents the cut levels as-is when Bullfrog last touched them, so many cannot be completed and are unbalanced, etc
- Placeholder mission briefings created for all Unguided campaign missions outlining what to do on each mission
- Unguided campaign missions missing in-game objectives now have objectives to complete
- Mission 65 - Massive Assault now uses the ALL000.BEN objectives rather than the final game ones. This fixes the incorrect group set for the first objective and also requires a weapon be recovered from the Eurocorp base to succeed.


Bug Fixes

- Item objective handling has been re-written. The original game code had many bugs in how items were handled and would fail on several of the cut Unguided levels as a result. These changes fix item collection objectives on those levels making it possible to complete them (e.g. Unguided Mission 67: Talks Ambush now works correctly)
- Issue 16: Item dupe by restarting missions fixed. This is the notorious bug in item handling that meant the player could end up with invisible explosives (and other so called "four pack" items) items in their inventories that could not be dropped when collected in a level and the level was restarted. This would make the penultimate Eurocorp level impossible to complete
- Issue 52: While it is possible to choose between two different Tension music tunes, only one ever plays fixed. The second MIDI tension song can now be chosen for the first time in-game. This was broken in the original DOS release of the game and only played the first track still when set
- Issue 46: Tank plays sound at map coords (0,0) when firing fixed. This was an original game bug that meant the sound intended to play when a tank turret rotated was only heard at one specific corner of the map, even when the tank was elsewhere in the level
- Issue 18: Number of scientist deaths is always double of the reported value fixed. This was an original game bug that meant research scientists death counts after completing a mission were wrong 
- Text fixes: Mission 43 is now correctly named "Deadly Harvest" in the English text files. The capital city of Bahrain is now called Al Manamah instead of Bahrain, as per the demos of the game.


Installer 

- Installer has been drastically overhauled and is new for this release
- Installer now supports all known Western versions of the game, including new support for the GoG download version and the German DOS CD
- Installer now supports the Japanese Windows version of the game. Note that only English and French language from this release work currently, Japanese language is not supported yet. The Taiwanese and Korean releases of the game aren't supported yet, if you own these, please get in touch!
- Installer now supports German language - GoG version includes these language files, otherwise you will need the specific German DOS CD release of the game. German language files were not included on the main European/USA disc.
- Note that ripping CD music is not supported and will error if you install from a mounted CD image of the game. You will either need a real CD of the game, or choose NOT to install the music in the installer, and instead rip the tracks manually yourself to a folder called "music" in the main game directory with the files names track_1.ogg, track_2.ogg and track_3.ogg.



Not sure about the PO files bit above, it's a cool feature but more source code than end user game related, so feels a bit wrong included it here, and also I'm not sure I expressed it well either. Obviously let me know if I missed anything important.
I also managed to fix the 100% installer bug, it actually just needed one Section edited, I just used a goto to skip the filecopy failure bit.

**Comment by mefistotelis** [comment]
Oh wow. I though those will be 6-10 lines, you really surprised me. But sure, that will do. Thank you.

**Comment by Moburma** [comment]
Well, it was obvious that people aren't aware of how much (or indeed any) work has been done on this project (mostly by your good self!), so I think it's worth going the extra mile to make it clear how good it is now and exactly what it can do compared to the original game.

**Comment by Moburma** [comment]
One annoying thing: I just downloaded the latest installer from the Github actions. It works fine, but Windows Defender has a false positive on it being a virus for some weird reason. Just what we needed.

**Comment by mefistotelis** [comment]
Not much we can do about that.

Well, most AVs have a way of reporting false positives, so that they can be manually verified and cleared in future versions. But I never bothered with that.

**Comment by geist22** [comment]
I've tried out the release version, though I normally prefer the simple flexibility of portable archives, here's some comments:

1. \QDATA\INTRO.SMK isn't needed here anymore, since this file is copied to \intro\, it's a duplicate now.
2. It couldn't load my old saves, I noticed the savegame location was changed from swars\qdata to user\appdata\roaming\ but I didn't know until after I've made a new save. Perhaps you might want to keep the old location enabled, or note that the location was changed. A new user wouldn't notice this, but I was a bit surprised when I've copied my saves to the usual directory and it didn't work. 

Other than that, I don't see anything to complain about. I've used the GOG version (which I've had to install again first, since I'm using two different copies, one for dosbox-staging and one for swars port).

**Comment by mefistotelis** [comment]
@Moburma are you willing to implement downloading map/level files and installing these?

An example release is already made in [swars-levels](https://github.com/swfans/swars-levels).
There's also a build systems which makes release ZIP automatically, so the file content will be consistent between releases.

What we need from the installer is:
1. Check if the ZIP file (with name hard-coded into the installer) exists in the installer folder. Optionally verify size or MD5 checksum as well (to avoid using half-downloaded files).
2. If the check fails, download the file from hard-coded link to a release within swars-levels to the installer folder.
3. While installing the port, unzip the levels file to the output folder rather than copying levels from original game.

**Comment by Moburma** [comment]
Yep, I will have a go. A bit busy with work this week but will see what I can do.
Do we want to make this mandatory or optional?

**Comment by mefistotelis** [comment]
Mandatory.

We should provide one experience.
If someone wants to diverge from it, that should be done by additional modifications after installation.
(or by installing manually rather than via the installer)

**Comment by Moburma** [comment]
I actually got this done already, but will be very busy the next couple of days so will return to it then.
I guess the idea is you'd have an include like the PRODUCT_VERSION define to dynamically add the swlevels zipfile name and MD5 during compilation? I've hardcoded those for now, but it seems the logical way to do it. Also it will need quite a few third party plugins as this is all non out-of-the-box functionality. Not sure the best way to do that using github actions yet, I'm just doing it locally for now.

At the moment I'm using these:
https://nsis.sourceforge.io/Inetc_plug-in - downloads
https://nsis.sourceforge.io/ZipDLL_plug-in - Zip file extract

And will likely add this for MD5 checking:
https://nsis.sourceforge.io/MD5_plugin

I'll upload a version once I've got the MD5 part worked out.

**Comment by mefistotelis** [comment]
> I guess the idea is you'd have an include like the PRODUCT_VERSION define to dynamically add the swlevels zipfile name and MD5 during compilation?

For now I wasn't thinking about moving the download link or md5 anywhere.
Sure this can be done, ie. we can use configure scripts to add these to `swars-config.nsh`, taking them from install instructions.
But for now, I see no issue with hard-coding.

> it will need quite a few third party plugins

Yeah, cannot avoid that. We just have to add installation of these plugins to our build environment. The action we use supports adding custom plugin folder:
https://github.com/joncloud/makensis-action

**Comment by Moburma** [comment]
Ok, finally had time to finish this off properly and submitted a change.
Needs the plugins I indicated above. Extracting the levels isn't very elegant, I found a better zip plugin that could extract only some files, but it wouldn't work - then I found it's an old plugin and doesn't have a unicode version that the newer installer expects, so I gave up on it. 
Anyway, this downloads the levels package, and compares the MD5 against what is stored in the installer. If it's bad it downloads again and checks once more. I didn't set it to keep trying after this as I hate it when things can get stuck in an infinite loop. If the second check is bad the installer aborts.

**Comment by mefistotelis** [comment]
Looks ok.

We should add the missing SCN files to swars-levels repo, then we can skip copying maps and levels from the original CD.
(they are supposed to be auto-generated, but I still don't have a tool for that)

**Comment by Moburma** [comment]
I've been working on a change to the installer. Now it will have a third option to update an existing install. This is handy, as it then gets the latest levels.zip etc for you. It also skips installing music etc. I've made it so it doesn't update config.ini and also skips copying rules.ini in this mode. I'm doing more testing now, but probably put a version up tomorrow.

**Comment by mefistotelis** [comment]
I recently updated the `util/install` script - it now copies specific files from the CD rather than taking whole directories.
We will need that as more and more files will come from our packages.

Maybe the NSIS installer could move to such selective copy as well.
You can check which files the `install` is copying in two ways: either just look into the script code, or use it to copy the CD to a specific folder and examine that folder.
(the 2nd solution may be easier in term of getting a flat list, because in the scripts file names are generated using `for()` loops.

I am also planning to soon update the `install` to use SFX package for sound banks.

For the NSIS installer, I would prefer to rely on you @Moburma for any substantial changes.

When switching to newer zip packages, we will now be updating both `util/install` and NSIS scripts.
(we could switch to templates so that the package versions and MD5 sums are in one place only.. but that maybe later)

**Comment by mefistotelis** [comment]
I just pushed the change to `install`. Yeah, sound banks are now per-language. I considered placing them in `language` folder, but it didn't felt right - so I finally settled on just including the language code in file name.

If no per-language file is found, there's a failsfafe to try the original `sound.dat` as well.

You do already have a list of all data files, for uninstall, right? Can NSIS support just having one list and reusing it for both functions? (the uninstall obviously removes more files, but some are shared).

> although to be honest surely it's easier to just copy the CD, and then overwrite with the downloaded files?

Yeah, that could still work. I don't have any overwhelming arguments for having a strict list. But I have some minor ones:
* If a CD is scratched, there's a higher chance of success when files are limited
* Over time, the amount of files used from CD will decrease. look at KeeperFX - it only gets like 3 files from original installation, and they're carefully selected to not be interesting for changes but still be crucial to starting the game
* If someone uses a damaged rip as a source of data, that will be noticed during installation preventing posts like "this doesn't work".

**Comment by Moburma** [comment]
> You do already have a list of all data files, for uninstall, right? Can NSIS support just having one list and reusing it for both functions? (the uninstall obviously removes more files, but some are shared).

I'm not sure about this, the simple answer is No, there is a second dedicated section just for uninstall files, so they have to be duplicated to an extent. However, I think it might be possible using e.g. an array. I will go away and look into this. 

> > although to be honest surely it's easier to just copy the CD, and then overwrite with the downloaded files?
> 
> Yeah, that could still work. I don't have any overwhelming arguments for having a strict list. But I have some minor ones:
> 
> * If a CD is scratched, there's a higher chance of success when files are limited
> * Over time, the amount of files used from CD will decrease. look at KeeperFX - it only gets like 3 files from original installation, and they're carefully selected to not be interesting for changes but still be crucial to starting the game
> * If someone uses a damaged rip as a source of data, that will be noticed during installation preventing posts like "this doesn't work".

To be honest I think it's unlikely to happen in reality - I would wager at least 70% of people installing are going to be through GOG (which is why I prioritised getting the installer to work with it), the remainder are mostly going to be iso images, and maybe 5% will be people using real discs. 

Anyway, I will do the sound.dat stuff in a minute as that's easy, then think about files/avoiding duplication/using an external file list.

**Comment by mefistotelis** [comment]
> maybe 5% will be people using real discs.

True. In the 10 years of this project existing I haven't seen a single mention of that from a real user.
I've seen people trying to use damaged rips though.
